test/Microsoft.Azure.Devices.Edge.Test.Common/config/EdgeConfigBuilder.cs

```diff
@@ -59,32 +59,48 @@ public IModuleConfigBuilder AddModule(string name, string image)
         // complexity as module names are assigned dynamically.
         public IEnumerable<EdgeConfiguration> BuildConfigurationStages()
         {
-            // Build all modules *except* edge agent
-            List<ModuleConfiguration> modules = this.moduleBuilders
-                .Where(b => b.Key != "$edgeAgent")
-                .Select(b => b.Value.Build())
-                .ToList();
-
-            // Build edge agent
-            modules.Insert(0, this.BuildEdgeAgent(modules));
-
             var config = new ConfigurationContent
             {
                 ModulesContent = new Dictionary<string, IDictionary<string, object>>()
             };
 
-            var moduleNames = new List<string>();
-            var moduleImages = new List<string>();
+            var moduleNames = new HashSet<string>();
+            var moduleImages = new HashSet<string>();
+
+            void UpdateConfiguration(ModuleConfiguration module)
+            {
+                moduleNames.Add(module.Name);
+                moduleImages.Add(module.Image);
+
+                if (module.DesiredProperties.Count != 0)
+                {
+                    config.ModulesContent[module.Name] = new Dictionary<string, object>
+                    {
+                        ["properties.desired"] = module.DesiredProperties
+                    };
+                }
+            }
+
+            // Edge agent is not optional; add if necessary
+            if (!this.moduleBuilders.ContainsKey(ModuleName.EdgeAgent))
+            {
+                this.AddEdgeAgent();
```

Comment:
Should this throw an error instead? Is it even possible to create a deployment w/o agent with this framework?

Reply:
This logic was already present; I just moved it into this function from `BuildEdgeAgent`. Yes, you can create a test that doesn't add edge agent, in which case it would get added here as a convenience. I don't think you could write such a test by using the higher-level `EdgeRuntime` class, but you could absolutely do it using the lower level classes directly.